vllm/model_executor/layers/linear.py

```diff
@@ -426,6 +445,11 @@ def weight_loader(self,
             start_idx = shard_id * shard_size
             loaded_weight = loaded_weight.narrow(output_dim, start_idx,
                                                  shard_size)
+        elif shard_dim is not None:
+            shard_size = loaded_weight.shape[shard_dim]
+            shard_index = ["q", "k", "v"].index(loaded_shard_id)
+            param_data = param_data.narrow(shard_dim, shard_index * shard_size,
+                                           shard_size)
```

Comment:
The above is what makes the codebooks work, and all codebook compression schemes can use this.

Reply:
tinyllama also uses GQA, as a smaller model to test https://huggingface.co/TinyLlama/TinyLlama-1.1B-Chat-v1.0

note the num_key_value_heads https://huggingface.co/TinyLlama/TinyLlama-1.1B-Chat-v1.0/blob/77e23968eed12d195bd46c519aa679cc22a27ddc/config.json#L16

Reply:
oh wait are they all the same size since this `elif` is only run when loading the codebooks? which are all the same size (regardless of head count)? If so this is very opaque when just reading the code

Reply:
Ah sorry I see `# shard_dim indicates fixed size concatenated at shard_id`

I think we should try to make this clearer (the name `shared_dim` doesn't really communicate this well)

Reply:
(especially since I had to go to `MergedColumnParallelLinear` to find this comment, at the very least we should repeat the comment in `QKVParallelLinear`)

Reply:
I flipped it to a bool `is_metadata` per Lucas's offline recommendation and sprinkled a few more comments in.